src/cmds/tx.py

```diff
@@ -0,0 +1,190 @@
+)
+from src.wallet.puzzles.p2_delegated_puzzle import (
+    puzzle_for_pk,
+    solution_for_conditions,
+)
+
+
+# TODO: From wallet/wallet.py: refactor
+def make_solution(primaries=None, min_time=0, me=None, consumed=None, fee=0):
+    assert fee >= 0
+    condition_list = []
+    if primaries:
+        for primary in primaries:
+            condition_list.append(
+                make_create_coin_condition(primary["puzzlehash"], primary["amount"])
+            )
+    if consumed:
+        for coin in consumed:
+            condition_list.append(make_assert_coin_consumed_condition(coin))
+    if min_time > 0:
+        condition_list.append(make_assert_time_exceeds_condition(min_time))
+    if me:
+        condition_list.append(make_assert_my_coin_id_condition(me["id"]))
+    if fee:
+        condition_list.append(make_assert_fee_condition(fee))
+    print(condition_list)
+    return solution_for_conditions(condition_list)
+
+
+class SpendRequest:
+    puzzle_hash: bytes32
+    amount: uint64
+
+    def __init__(self, ph, amt):
+        self.puzzle_hash = ph
+        self.amount = amt
+
+
+def create_unsigned_transaction(
+    pubkey: bytes32,
+    coins: Set[Coin] = None,  # Set[CoinWithPuzzle] = None,
+    spend_requests: Set[SpendRequest] = None,
+    validate=True,
+) -> List[CoinSolution]:
+    """
+    Generates a unsigned transaction in form of List(Puzzle, Solutions)
+    """
+
+    if coins is None or len(coins) < 1:
+        raise (ValueError("tx create requires one or more input_coins"))
+    assert len(coins) > 0
+
+    input_value = sum([coin.amount for coin in coins])
+    sent_value = sum([req.amount for req in spend_requests])
+    if validate and sent_value != input_value:
+        raise (
+            ValueError(
+                f"input amounts ({input_value}) do not equal outputs ({sent_value})"
+            )
+        )
+
+    spends: List[CoinSolution] = []
+
+    # We treat the first coin as the origin
+    # For simplicity, only the origin coin creates outputs
+    origin = coins[0]
+    outputs = []
+
+    # Eventually, we will support specifying the puzzle directly in the
+    # input to create_unsigned_transaction (viaCoinWithPuzzle).
+    # For now, we specify a pubkey, and use the "standard transaction"
+    puzzle = puzzle_for_pk(pubkey)
+
+    for request in spend_requests:
+        outputs.append({"puzzlehash": request.puzzle_hash, "amount": request.amount})
+
+    solution = make_solution(primaries=outputs, fee=0)
+    puzzle_solution_pair = Program.to([puzzle, solution])
+    spends.append(CoinSolution(origin, puzzle_solution_pair))
+
+    # Create nil solutions for the other input coins
+    for coin in coins[1:]:
+        print(f"processing coin {coin}")
+        solution = make_solution()
+        puzzle_solution_pair = Program.to([coin.puzzle, solution])
+        spends.append(CoinSolution(coin, puzzle_solution_pair))
+
+    return spends
+
+
+def create_unsigned_tx_from_json(json_tx):
```

Comment:
I don't know if it's appropriate, but the Streamable class also has built in Json serialization, which is used for the http/websockets RPC. So things like FullBlock can be easily converted into Json. you might be able to avoid writing any additional JSON code.

Reply:
That also includes an exception for Program, which just serializes it to hex. So it will handle your TODO below for CoinSolution.